src/wutil/gettext.rs

```diff
@@ -1,136 +1,253 @@
+                }
+                None => {
+                    if let Some(catalog) = load_catalog_for_language(&language) {
+                        let message = catalog.gettext(&key_string).to_owned();
+                        catalogs.insert(language, catalog);
+                        return Box::leak(str2wcstring(message.as_bytes()).into_boxed_utfstr());
+                    }
+                }
+            };
+        }
+        Box::leak(key.to_owned().into_boxed_utfstr())
     }
 }
-#[cfg(not(gettext))]
-mod internal {
-    use libc::c_char;
-    use std::ffi::CStr;
-    pub fn fish_gettext(msgid: &CStr) -> *const c_char {
-        msgid.as_ptr()
-    }
-    pub fn fish_bindtextdomain(_domainname: &CStr, _dirname: &CStr) -> *mut c_char {
-        std::ptr::null_mut()
+
+#[cfg(not(feature = "embed-data"))]
+mod wgettext_impl {
+    use super::MaybeStatic;
+    use crate::common::{charptr2wcstring, truncate_at_nul, wcs2zstring, PACKAGE_NAME};
+    use crate::wchar::prelude::*;
+    use errno::{errno, set_errno};
+    use once_cell::sync::{Lazy, OnceCell};
+    use std::collections::HashMap;
+    use std::ffi::CString;
+    use std::sync::Mutex;
+
+    #[cfg(gettext)]
+    mod internal {
+        use libc::c_char;
+        use std::ffi::CStr;
+        extern "C" {
+            fn gettext(msgid: *const c_char) -> *mut c_char;
+            fn bindtextdomain(domainname: *const c_char, dirname: *const c_char) -> *mut c_char;
+            fn textdomain(domainname: *const c_char) -> *mut c_char;
+        }
+        pub fn fish_gettext(msgid: &CStr) -> *const c_char {
+            unsafe { gettext(msgid.as_ptr()) }
+        }
+        pub fn fish_bindtextdomain(domainname: &CStr, dirname: &CStr) -> *mut c_char {
+            unsafe { bindtextdomain(domainname.as_ptr(), dirname.as_ptr()) }
+        }
+        pub fn fish_textdomain(domainname: &CStr) -> *mut c_char {
+            unsafe { textdomain(domainname.as_ptr()) }
+        }
     }
-    pub fn fish_textdomain(_domainname: &CStr) -> *mut c_char {
-        std::ptr::null_mut()
+    #[cfg(not(gettext))]
+    mod internal {
+        use libc::c_char;
+        use std::ffi::CStr;
+        pub fn fish_gettext(msgid: &CStr) -> *const c_char {
+            msgid.as_ptr()
+        }
+        pub fn fish_bindtextdomain(_domainname: &CStr, _dirname: &CStr) -> *mut c_char {
+            std::ptr::null_mut()
+        }
+        pub fn fish_textdomain(_domainname: &CStr) -> *mut c_char {
+            std::ptr::null_mut()
+        }
     }
-}
 
-use internal::*;
+    use internal::*;
 
-// Really init wgettext.
-fn wgettext_really_init() {
-    let Some(ref localepath) = CONFIG_PATHS.locale else {
-        return;
-    };
-    let package_name = CString::new(PACKAGE_NAME).unwrap();
-    // This contains `datadir`; which when replaced to make the binary relocatable,
-    // causes null bytes at the end of the string.
-    let localedir = CString::new(localepath.display().to_string()).unwrap();
```

Comment:
Is it correct that `CONFIG_PATHS.locale` should be None when we have embedded the `*.mo` files?
If yes, we probably need to change the logic in `src/env/config_paths.rs` accordingly.

Reply:
ah yes, that's already what's happening.. except if the path to fish ends in `bin/fish`,
or if `$(status fish-path)/../../share/install` exists.. though that one is purely a legacy thing (since we have embedding now), so we can probably remove the "install" fallback.. not yet sure)